src/main/java/org/jenkinsci/plugins/workflow/multibranch/ReadTrustedStep.java

```diff
@@ -0,0 +1,188 @@
+            }
+            SCMSource scmSource = ((WorkflowMultiBranchProject) parent).getSCMSource(branch.getSourceId());
+            if (scmSource == null) {
+                throw new IllegalStateException(branch.getSourceId() + " not found");
+            }
+            SCMHead head = branch.getHead();
+            SCMRevision tip;
+            SCMRevisionAction action = build.getAction(SCMRevisionAction.class);
+            if (action != null) {
+                tip = action.getRevision();
+            } else {
+                tip = scmSource.fetch(head, listener);
+                if (tip == null) {
+                    throw new AbortException("Could not determine exact tip revision of " + branch.getName());
+                }
+                build.addAction(new SCMRevisionAction(tip));
+            }
+            SCMRevision trusted = scmSource.getTrustedRevision(tip, listener);
+            // Adapted from CpsScmFlowDefinition:
+            Node node = Jenkins.getActiveInstance();
+            FilePath dir;
+            if (job instanceof TopLevelItem) {
+                FilePath baseWorkspace = node.getWorkspaceFor((TopLevelItem) job);
+                if (baseWorkspace == null) {
+                    throw new AbortException(node.getDisplayName() + " may be offline");
+                }
+                dir = getFilePathWithSuffix(baseWorkspace);
+            } else { // should not happen, but just in case:
+                throw new IllegalStateException(job + " was not top level");
+            }
+            FilePath file = dir.child(step.path);
+            if (!file.absolutize().getRemote().replace('\\', '/').startsWith(dir.absolutize().getRemote().replace('\\', '/') + '/')) { // TODO JENKINS-26838
+                throw new IOException(file + " is not inside " + dir);
+            }
+            Computer computer = node.toComputer();
+            if (computer == null) {
+                throw new IOException(node.getDisplayName() + " may be offline");
+            }
+            WorkspaceList.Lease lease = computer.getWorkspaceList().acquire(dir);
+            try {
+                String untrustedFile = null;
+                if (!tip.equals(trusted)) {
+                    SCMStep delegate = new GenericSCMStep(scmSource.build(head, tip));
+                    delegate.setPoll(false);
+                    delegate.setChangelog(false);
+                    delegate.checkout(build, dir, listener, node.createLauncher(listener));
+                    if (!file.exists()) {
+                        throw new AbortException(file + " not found");
+                    }
+                    untrustedFile = file.readToString();
+                }
+                SCMStep delegate = new GenericSCMStep(scmSource.build(head, trusted));
+                delegate.setPoll(true);
+                delegate.setChangelog(true);
+                delegate.checkout(build, dir, listener, node.createLauncher(listener));
```

Comment:
Two checkouts... maybe the `SCMSource` should be used (new APIs needed I think). I would be much more efficient for API accesible SCMs.